team/levels.tex

```diff
@@ -0,0 +1,105 @@
+    Understands usage of compilers, interpreters, build tools used at the organisation. \\
+  \hline
+    2 & & &
+    Applies best practices industry security guidelines, like setting up strong TLS, can pick strong authentication and authorization mechanisms.&
+    &
+    Has more granular understanding of the networking design, for example can reason about using GRPC vs HTTPS-JSON and their networking and scalability trade-offs.
+    Can do the same for UDP vs TCP and lower level protocols.&
+    Can reason about performance implications and risks of using synchronisation primitives, understands granularity of locking, can debug and troubleshoot memory and synchronisation issues.&
+    Understands tools and compilers as advanced user - for example, can refactor Makefile to make it more efficient and parallel or select most optimal compilation flags of a complier for Linux and MacOS. \\
+  \hline
+    3 &
+    Uses advanced data structures and algorithms, such as consensus, gossip protocols to implement key product features.&
+    Collaborates with the team to scope requirements, based on good understanding of existing longer term product vision and estimates of the system design of a feature of a product.&
+    Understands security aspects of protocols on a deeper level, for example can explain differences between TLS 1.2 and 1.3 and security implications.
+    Understands common attack vectors for server side or client side applications and can build secure systems that will pass quality security audit that will uncover no critical system design errors.&
+    &
+    Understands scalability and resilience aspects of practical implementation of systems leveraging networking.
+    Can write a fast, scalable networking servers and troubleshoot common networking issues in the servers and client design, such as connection lifecycle and pooling, backpressure and other aspects of the design. &
+    Not only can write data-race and deadlock free code, but can implement most safe and concurrent and parallel systems using minimum amount of shared state, granular locking and easy to read and extend and troubleshoot system.
+    Senior engineer usually can not find any faults or provide any suggestions in the systems design submitted by the Level 3 engineer.
+    & \\
+  \hline
+    4 &
+    Leads implementation of products and standalone systems using advanced data structures and algorithms, such as upgrade/install distributed framework.&
+    Leads the implementation of the isolated feature/improvement that measurably and significantly impacts business outcomes from gathering requirements to getting to the market stage.&
+    Can implement production quality existing novel crytpography and security systems considering tradeoffs. For example can build eBPF driven security into the system from scratch that will pass quality external security audit that will uncover no critical system design errors.&
```

Comment:
```suggestion
    Can implement production quality novel cryptographic and security systems considering the trade-offs. For example, can implement strong security support for the system with eBPF from scratch.&
```
Let me know if I got the context of `existing` wrong - I thought it was out of place there. Also I thought the audit part of redundant given the security is strong.